### PR TITLE
CY-1244 Plugins install: replace local lock mechanism with DB lock,

### PR DESCRIPTION
**Pull request by adigrabow**
and wait for Syncthing to sync files on all Managers if cluser

### CODE REVIEW

File: cloudify_agent/api/plugins/installer.py [hunk 1]
```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
...
+        :return:
+        """
+        client = get_rest_client()
+        client.plugins.finish_installation(plugin_id)
+
+    @staticmethod
+    def _is_plugin_installing(plugin_id):
+        client = get_rest_client()
+        plugin = client.plugins.get(plugin_id)
+        if 'installing-' in plugin.archive_name:
```
Comment (tyacbovi) [inline cloudify_agent/api/plugins/installer.py]: I think it's better to put this logic in a shared location, so does the agent will not have to impl this again

File: cloudify_agent/api/plugins/installer.py [hunk 2]
```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
...
+            return True
```
Comment (tyacbovi) [inline cloudify_agent/api/plugins/installer.py]: This can be combined to one liner

File: cloudify_agent/api/plugins/installer.py [hunk 3]
```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
...
+    def _wait_for_syncthing():
+        # This file exists only when the Syncthing service is running
+        if os.path.exists(syncthing_utils.config_path):
+            while not syncthing_utils.mgmtworker_is_syncing_complete():
+                time.sleep(1)
+
+    @staticmethod
+    def _update_plugin_status(plugin_id):
+        """
+        Now that both the installation and Syncthing has ended the `installing`
```
Comment (tyacbovi) [inline cloudify_agent/api/plugins/installer.py]: I will reword it to 'Completing the plugin installation process after installing it on relevant locations and syncing completion'

File: cloudify_agent/api/plugins/installer.py [hunk 4]
```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
...
-                            'does not match the ID of the plugin currently '
-                            'on the manager. [existing: {0}, new: {1}]'
-                            .format(existing_plugin_id,
-                                    managed_plugin.id))
-                else:
+
+        # create_deployment_env should wait for the install_plugin wf to finish
+        while (self._is_plugin_installing(managed_plugin.id) and
+               deployment_id != SYSTEM_DEPLOYMENT):
+            time.sleep(1)
```
Comment (tyacbovi) [inline cloudify_agent/api/plugins/installer.py]: Better const the one second sleep, and maybe add the explanation to the connection to the syncthing loop

Comment (barucoh) [inline cloudify_agent/api/plugins/installer.py]: magic number, perhaps export to a constant?
Also, it's worth checking if checking the DB every 1 second is a good idea.

File: cloudify_agent/api/plugins/installer.py [hunk 5]
```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
...
+                raise NonRecoverableError('Failed installing managed '
+                                          'plugin: {0} [{1}][{2}]'
+                                          .format(managed_plugin.id,
+                                                  plugin, e)), None, tb
+
+    @staticmethod
+    def _wait_for_syncthing():
+        # This file exists only when the Syncthing service is running
+        if os.path.exists(syncthing_utils.config_path):
+            while not syncthing_utils.mgmtworker_is_syncing_complete():
```
Comment (barucoh) [inline cloudify_agent/api/plugins/installer.py]: Don't forget to update the function name here

File: cloudify_agent/api/plugins/installer.py [hunk 6]
```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
...
+                time.sleep(1)
```
Comment (barucoh) [inline cloudify_agent/api/plugins/installer.py]: magic number, same as before

File: cloudify_agent/api/plugins/installer.py (hunk 2 above)
Comment (barucoh) [inline cloudify_agent/api/plugins/installer.py]: can be replaced with
`return plugin.archive_name.starts_with(INSTALL_PREFIX)`

File: cloudify_agent/tests/api/plugins/test_installer.py [hunk 7]
```diff
@@ -503,7 +511,46 @@ def list(self, **kwargs):
...
+                         format(plugin_id))
+            super(ConcurrentMockPlugins, self).finish_installation(plugin_id)
+
 
 class MockClient(object):
     def __init__(self, plugins, download_path=None):
         self.plugins = MockPlugins(plugins, download_path=download_path)
+
+
+class MockConcurrentClinet(object):
```
Comment (tehasdf) [inline cloudify_agent/tests/api/plugins/test_installer.py]: s/Clinet/Client/

File: cloudify_agent/api/plugins/installer.py [hunk 8]
```diff
@@ -135,67 +142,107 @@ def _install_managed_plugin(self,
...
+        :param plugin_id:
+        """
+        client = get_rest_client()
+        client.plugins.finish_installation(plugin_id)
+
+    @staticmethod
+    def _is_plugin_installing(plugin_id):
+        client = get_rest_client()
+        plugin = client.plugins.get(plugin_id)
+        return True if plugin.archive_name.startswith(INSTALLING_PREFIX)\
```
Comment (tehasdf) [inline cloudify_agent/api/plugins/installer.py]: this is just `return plugin.archive_name.startswith(INSTALLING_PREFIX)` :)

File: cloudify_agent/api/plugins/installer.py [hunk 9]
```diff
@@ -135,67 +142,107 @@ def _install_managed_plugin(self,
...
+                self._update_plugin_status(managed_plugin.id)
+            except Exception as e:
+                tpe, value, tb = sys.exc_info()
+                raise NonRecoverableError('Failed installing managed '
+                                          'plugin: {0} [{1}][{2}]'
+                                          .format(managed_plugin.id,
+                                                  plugin, e)), None, tb
+
+    @staticmethod
+    def _wait_for_syncthing():
```
Comment (tehasdf) [inline cloudify_agent/api/plugins/installer.py]: this whole function has to do with waiting for replication so i'd put it in premium

and then this module wouldn't need to know anything about the replication impl (now it needs to know the config path and that it needs to poll)

File: cloudify_agent/api/plugins/installer.py [hunk 10]
```diff
@@ -135,67 +142,107 @@ def _install_managed_plugin(self,
...
-            if lock:
-                lock.release()
+                self._wagon_install(plugin=managed_plugin, args=args)
+                shutil.move(tmp_plugin_dir, dst_dir)
+                with open(os.path.join(dst_dir, 'plugin.id'), 'w') as f:
+                    f.write(managed_plugin.id)
+                # Wait for Syncthing to sync plugin files on all managers
+                if syncthing_utils:
+                    self._wait_for_syncthing()
+                self._update_plugin_status(managed_plugin.id)
```
Comment (tehasdf) [inline cloudify_agent/api/plugins/installer.py]: if plugin installation fails we don't set anything, unfortunate
